src/lib.rs

```diff
@@ -0,0 +1,389 @@
+    /// assert!(matcher.matches("a b b").is_err());
+    /// ```
+    pub fn name_matcher(mut self, matcher: Option<(Regex, Regex)>) -> Self {
+        self.options.name_matcher = matcher;
+        self
+    }
+
+    #[doc(hidden)]
+    pub fn ignore_leading_whitespace(mut self, yes: bool) -> Self {
+        self.options.ignore_leading_whitespace = yes;
+        self
+    }
+
+    #[doc(hidden)]
+    pub fn ignore_trailing_whitespace(mut self, yes: bool) -> Self {
+        self.options.ignore_trailing_whitespace = yes;
+        self
+    }
+
+    /// Turn this `FMBuilder` into a `FMatcher`.
+    pub fn build(self) -> Result<FMatcher<'a>, Box<dyn Error>> {
+        Ok(FMatcher {
+            ptn: self.ptn,
+            options: self.options,
+        })
+    }
+}
+
+/// The fuzzy matcher.
+pub struct FMatcher<'a> {
+    ptn: &'a str,
+    options: FMOptions,
+}
+
+impl<'a> FMatcher<'a> {
+    /// A convenience method that automatically builds a pattern for you using `FMBuilder`'s
+    /// default options.
+    pub fn new(ptn: &'a str) -> Result<FMatcher, Box<dyn Error>> {
+        FMBuilder::new(ptn)?.build()
+    }
+
+    /// Does this fuzzy matcher match `text`? Returns `Ok(())` on success or `Err(usize)` on error
+    /// where the `usize` is the line number of the first line in `text` where the match failed.
+    pub fn matches(&self, text: &str) -> Result<(), usize> {
+        let mut names = HashMap::new();
+        let mut ptn_lines = self.ptn.lines();
+        let mut ptnl = ptn_lines.next();
+        let mut text_lines = text.lines();
+        let mut textl = text_lines.next();
+        let mut text_lines_off = 1;
+        loop {
+            match (ptnl, textl) {
+                (Some(x), Some(y)) => {
+                    if x.trim() == WILDCARD {
+                        ptnl = ptn_lines.next();
+                        match ptnl {
+                            Some(x) => {
+                                while let Some(y) = textl {
+                                    if self.match_line(&mut names, x, y) {
+                                        break;
+                                    }
+                                    textl = text_lines.next();
+                                    text_lines_off += 1;
+                                }
+                                text_lines_off -= 1;
+                            }
+                            None => return Ok(()),
+                        }
+                    } else if self.match_line(&mut names, x, y) {
+                        ptnl = ptn_lines.next();
+                        textl = text_lines.next();
+                        text_lines_off += 1;
+                    } else {
+                        return Err(text_lines_off);
+                    }
+                }
+                (None, None) => return Ok(()),
+                (Some(x), None) if x.trim() == WILDCARD => return Ok(()),
+                (Some(_), None) => return Err(text_lines_off),
```

Comment:
You can merge the latter two cases I think.

Reply:
? They're very different -- look at the RHS of the `=>`: one returns `Ok`, the other `Err`.

Reply:
I mean the last two lines of the match, sorry.

```
match ... {
    (Some(_), None) | (None, Some(_)) => return Err(text_lines_off),
```

Reply:
Good point! https://github.com/softdevteam/fm/pull/1/commits/a538c0425b4614fd152bc1ed32ac4cbe01211ff6